src/components/MainNav/AppOrderProvider.js

```diff
@@ -0,0 +1,166 @@
+import { createContext, useContext, useMemo } from 'react';
+import { useLocation } from 'react-router-dom';
+import { useIntl } from 'react-intl';
+import { useQuery } from 'react-query';
+
+import { useStripes } from '../../StripesContext';
+import { useModules } from '../../ModulesContext';
+import { LastVisitedContext } from '../LastVisited';
+import usePreferences from '../../hooks/usePreferences';
+import { packageName } from '../../constants';
+import settingsIcon from './settings.svg';
+
+
+const APPORDER_PREF_NAME = 'user-main-nav-order';
+const APPORDER_PREF_SCOPE = 'stripes-core.prefs.manage';
+
+export const AppOrderContext = createContext({
+  isLoading: true,
+  listOrder: [],
+  apps: [],
+  updateList: () => {},
+  reset: () => {},
+});
+
+export const useAppOrderContext = () => {
+  return useContext(AppOrderContext);
+};
+
+function getProvisionedApps(appModules, stripes, pathname, lastVisited, formatMessage) {
+  const apps = appModules.map((entry) => {
+    const name = entry.module.replace(packageName.PACKAGE_SCOPE_REGEX, '');
+    const perm = `module.${name}.enabled`;
+
+    if (!stripes.hasPerm(perm)) {
+      return null;
+    }
+
+    const id = `clickable-${name}-module`;
+
+    const pathRoot = pathname.split('/')[1];
+    const entryRoot = entry.route.split('/')[1];
+    const active = pathRoot === entryRoot;
+
+    const last = lastVisited[name];
+    const home = entry.home || entry.route;
+    const href = (active || !last) ? home : lastVisited[name];
+
+    return {
+      id,
+      href,
+      active,
+      name,
+      ...entry,
+    };
+  }).filter(app => app);
+
+  /**
+   * Add Settings to apps array manually
+   * until Settings becomes a standalone app
+   */
+
+  if (stripes.hasPerm('settings.enabled')) {
+    apps.push({
+      displayName: formatMessage({ id: 'stripes-core.settings' }),
+      name: 'settings',
+      id: 'clickable-settings',
+      href: lastVisited.x_settings || '/settings',
+      active: pathname.startsWith('/settings'),
+      description: 'FOLIO settings',
+      iconData: {
+        src: settingsIcon,
+        alt: 'Tenant Settings',
+        title: 'Settings',
+      },
+      route: '/settings'
+    });
+  }
+  return apps;
+  // return sortby(apps, app => app.displayName.toLowerCase());
+}
+
+export const AppOrderProvider = ({ children }) => {
+  const { lastVisited } = useContext(LastVisitedContext);
+  const { app } = useModules();
+  const stripes = useStripes();
+  const { pathname } = useLocation();
+  const { formatMessage } = useIntl();
+  const { getPreference, setPreference, removePreference } = usePreferences();
+
+  const { data: userAppList, isLoading, refetch: refetchAppList } = useQuery('UserPrefsAppListing', () => {
+    return getPreference({ key: APPORDER_PREF_NAME, scope: APPORDER_PREF_SCOPE });
```

Comment:
Maybe use an expression here rather than a block that does nothing but return the expression?